### PR TITLE
WIP: add memory manegement design doc

### PR DESCRIPTION
**Pull request by gangliao**

### CODE REVIEW

File: memory/cpu/system_allocator.cc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    return system_allocators.size();
+}
+
+void SystemAllocator::init() {
+    assert(system_allocators.empty());
+
+    // make sure no copies occur
+    system_allocators.reserve(2);
+
+    // add the pinned allocator
```
Comment (gangliao) [inline memory/cpu/system_allocator.cc]: How to manually configured to use `PinnedAllocator ` ?

In Majel, it records system environment variables as follows:

```c++
// add the pinned allocator
    if (support::get_knob_value("MAJEL_USE_PINNED_ALLOCATOR", false)) {
        system_allocators.push_back(std::unique_ptr<Allocator>(new PinnedAllocator));
    }

    // add the default allocator
    system_allocators.push_back(std::unique_ptr<Allocator>(new DefaultAllocator));
```

If we use gflag to replace it, can we easily switch to the different allocator in Python?

File: memory/README.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+## Region-based Heterogeneous Memory Management
+
+With the recent explosion of deep learning, entire industries substantial concerns around how to build
```
Comment (wangkuiyi) [inline memory/README.md]: This paragraph reads like a scientific paper that tries to advertise the importance of the research work. :D  I guess we don't need it.

Indeed, I think we need a Wikipedia link https://en.wikipedia.org/wiki/Region-based_memory_management to explain the concept *Region-based memory management*.

**Comment by wangkuiyi** [comment]
I think these files should be in `/paddle/memory` instead of `/memory`?

**Comment by wangkuiyi** [comment]
Just record here:

TensorFlows' [CUDAContext](https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/stream_executor/cuda/cuda_driver.h#L482) looks similar to Majel's Place.

**Comment by wangkuiyi** [comment]
I know [how Caffe2 allocates memory](https://github.com/PaddlePaddle/Paddle/wiki/Caffe2::Tensor), and I reviewed Majel's memory allocation as well.  These work helped me understand this design. However, I was not able to figure out how `tensorflow::Tensor` allocate GPU memory: https://github.com/PaddlePaddle/Paddle/issues/2493

File: paddle/memory/cpu/detail/pinning-inl.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#else
+#include <sys/mman.h>
+
+#endif
+
+namespace paddle {
+namespace memory {
+namespace cpu {
+
+inline void pin_memory(void* address, size_t size) {
```
Comment (wangkuiyi) [inline paddle/memory/cpu/detail/pinning-inl.h]: I'd recommend to merge pinning-inl.h and pinning.h, as both are very short. I have some more detailed ideas in https://github.com/PaddlePaddle/Paddle/issues/2505

File: paddle/memory/malloc.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <paddle/majel/place.h>
+
+namespace paddle {
+namespace memory {
+
+void init();
+void shutdown();
+
+void* malloc(majel::Place place, size_t size);
```
Comment (wangkuiyi) [inline paddle/memory/malloc.h]: How about we add the following two functions here:

```cpp
malloc_pinned_and_mapped(size_t host_memory_size);
malloc_pinned(size_t host_memory_size);
```

so we can write 

1. GpuAllocator, which corresponds to Majels' `cpu/SystemAllocator`, 
1. GPUAllocator, which corresponds to Majel's `gpu/SystemAllocaotr`, and
1. BuddyAllocator, which corresponds to Majels' `detail/BuddyAllocator`, which fail backs to either GPUAllocator or CPUAllocaotr.

**Comment by wangkuiyi** [comment]
@gangliao In https://github.com/PaddlePaddle/Paddle/pull/2513, I am creating

- `paddle/platform`, which includes Place, and
- `paddle/framework`, which includes Dim/DDim and will include Tensor.

`paddle/framework` depends on `paddle/platform`.

It seems that this PR creates `paddle/memory`, which should depend also on `paddle/platform` (for Place). What do you think about this?

File: memory/cpu/detail/pinning-inl.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+inline void unpin_memory(void* address, size_t size) {
+    #if _WIN32
+
+    #else
+        munlock(address, size);
+    #endif
+}
+
+} /* cpu */
```
Comment (gangliao) [inline memory/cpu/detail/pinning-inl.h]: done

File: memory/README.md (hunk 2 above)
Comment (gangliao) [inline memory/README.md]: done

File: memory/cpu/system_allocator.cc (hunk 1 above)
Comment (gangliao) [inline memory/cpu/system_allocator.cc]: using flags..

File: paddle/memory/cpu/detail/pinning-inl.h (hunk 3 above)
Comment (gangliao) [inline paddle/memory/cpu/detail/pinning-inl.h]: done

**Review by wangkuiyi — COMMENTED** [review]
For your reference,  I wrote a design (not implementation at all) in https://github.com/PaddlePaddle/Paddle/pull/2552.

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献文档。由于文档已迁移至[FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo，因此关闭您的PR，欢迎您向[FluidDoc](https://github.com/PaddlePaddle/FluidDoc) Repo贡献文档。
Thanks for contributing to PaddlePaddle! Since documents have been moved to [FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo, we close this PR. Welcome to contribute to [FluidDoc](https://github.com/PaddlePaddle/FluidDoc) repo.

**Pull request closed without merging**